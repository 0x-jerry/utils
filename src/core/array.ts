import { isFn, isPrimitive } from '../is'
import { Arrayable, PrimitiveType } from '../types'

/**
 * Ensure to return a list of element T.
 *
 * @param arr
 * @returns
 */
export const ensureArray = <T>(arr?: Arrayable<T>): T[] => {
  if (arr == null) {
    return []
  }

  return Array.isArray(arr) ? arr : [arr]
}

/**
 * @deprecated use {@link ensureArray} instead of.
 */
export const toArray = ensureArray

export const remove = <T>(arr: T[], predict: T | ((item: T) => boolean)): number => {
  const idx = isFn(predict) ? arr.findIndex(predict) : arr.indexOf(predict)

  if (idx >= 0) arr.splice(idx, 1)

  return idx
}

type GroupResult<T, U> = U extends PrimitiveType
  ? Record<string | number | symbol, T[]>
  : Map<U, T[]>

/**
 *
 * @experiment
 * @param arr
 * @param callbackFn
 * @returns
 */
export function group<T, U>(arr: T[], callbackFn: (item: T) => U): GroupResult<T, U> {
  let record: any = null

  let recordIsMap = false

  function initRecord(value: any) {
    if (isPrimitive(value)) {
      record = {}
    } else {
      recordIsMap = true
      record = new Map()
    }
  }

  for (const item of arr) {
    const value = callbackFn(item)

    if (!record) initRecord(value)

    const exist = recordIsMap ? record.get(value) : record[value]

    if (exist) {
      exist.push(item)
    } else {
      const g = [item]

      if (recordIsMap) {
        record.set(value, g)
      } else {
        record[value] = g
      }
    }
  }

  return record || new Map()
}
