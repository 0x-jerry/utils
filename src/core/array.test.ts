import { group, remove, toArray } from './array'

describe('toArray', () => {
  it('array', () => {
    expect(toArray([1])).toEqual([1])
  })

  it('primary value', () => {
    expect(toArray(1)).toEqual([1])

    expect(toArray('')).toEqual([''])
  })

  it('2 dimension', () => {
    expect(toArray([[1], [2]])).toEqual([[1], [2]])
  })
})

describe('remove', () => {
  it('should remove exists item', () => {
    const item = [1, 2, { a: 1 }]

    expect(remove(item, 2)).toBe(1)
    expect(item).toEqual([1, { a: 1 }])
    expect(remove(item, (o: any) => o.a === 1)).toBe(1)
    expect(item).toEqual([1])
  })

  it('should return -1', () => {
    const item = [1, 2]

    expect(remove(item, 3)).toBe(-1)
    expect(remove(item, (o) => o === 3)).toBe(-1)
    expect(item).toEqual([1, 2])
  })
})

describe('group', () => {
  it('should group by string', () => {
    const items = [
      { type: 1, name: '1' },
      { type: 1, name: '2' },
      { type: 2, name: '3' },
    ]

    const result = group(items, (item) => item.type)

    expect(result).toMatchSnapshot()
  })

  it('should group by object', () => {
    const keys = [{ key: 1 }, { key: 2 }]

    const items = [
      { type: keys[0], name: '1' },
      { type: keys[0], name: '2' },
      { type: keys[1], name: '3' },
    ]

    const result = group(items, (item) => item.type)

    expect(result).toMatchSnapshot()
  })
})
