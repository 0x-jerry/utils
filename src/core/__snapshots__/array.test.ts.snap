// Vitest Snapshot v1

exports[`group > should group by object 1`] = `
Map {
  {
    "key": 1,
  } => [
    {
      "name": "1",
      "type": {
        "key": 1,
      },
    },
    {
      "name": "2",
      "type": {
        "key": 1,
      },
    },
  ],
  {
    "key": 2,
  } => [
    {
      "name": "3",
      "type": {
        "key": 2,
      },
    },
  ],
}
`;

exports[`group > should group by string 1`] = `
{
  "1": [
    {
      "name": "1",
      "type": 1,
    },
    {
      "name": "2",
      "type": 1,
    },
  ],
  "2": [
    {
      "name": "3",
      "type": 2,
    },
  ],
}
`;
